File: Tests/FakeServices/FakeSignalingChannel.m
//
//  FakeSignalingChannel.m
//  RTCDemo
//
//  Created by Anton Sokolchenko on 1/28/16.
//  Copyright © 2016 anton. All rights reserved.
//

#import "FakeSignalingChannel.h"
#import "SVSignalingChannelProtocol.h"


@implementation FakeSignalingChannel

@synthesize user = _user;

@synthesize delegate;

- (instancetype)init {
	self = [super init];
	if (self) {
		self.shouldSendMessagesSuccessfully = YES;
	}
	return self;
}

- (void)connectWithUser:(SVUser *)user completion:(void (^)(NSError * _Nullable))completion {
	self.state = SVSignalingChannelState.open;
	
	self.user = user;
	
	self.state = SVSignalingChannelState.established;
	
	if (completion) {
		completion(nil);
	}
}

- (void)notifyDelegateWithCurrentState {
	if (self.delegate) {
		if ([self.delegate respondsToSelector:@selector(channel:didChangeState:)]) {
			[self.delegate channel:self didChangeState:self.state];
		}
	}
}

- (void)setState:(NSString *)state {
	if (![_state isEqualToString:state]) {
		_state = state;
		[self notifyDelegateWithCurrentState];
	}
}


- (BOOL)isConnected {
	return YES;
}

- (void)sendMessage:(__kindof SVSignalingMessage *)message toUser:(SVUser *)user completion:(void (^)(NSError * _Nullable))completion {
	if (completion) {
		if (self.shouldSendMessagesSuccessfully) {
			completion(nil);
		} else {
			completion([[NSError alloc] init]);
		}
		
	}
}

- (void)disconnectWithCompletion:(void (^)(NSError * _Nullable))completion {
	if (completion) {
		completion(nil);
	}
}

@end

File: Tests/FakeServices/FakeSignalingChannel.swift
//
//  FakeSignalingChannel.swift
//  QBRTCDemo
//
//  Created by Anton Sokolchenko on 09.12.16.
//  Copyright © 2016 Anton Sokolchenko. All rights reserved.
//

import Foundation

#if QBRTCDemo_s
	@testable
	import QBRTCDemo_s
#elseif QBRTCDemo
	@testable
	import QBRTCDemo
#endif

class FakeSignalingChannel: NSObject {
	
	/// Should send messages successfully or with error
	var shouldSensMessagesSuccessfully = true
	
	var state = SignalingChannelState.open
	var user: SVUser?
	
	var observers = MulticastDelegate<SignalingChannelObserver>()
}

extension FakeSignalingChannel: SignalingChannelProtocol {
	
	func addObserver(_ observer: SignalingChannelObserver) {
		observers += observer
	}
	
	func connectWithUser(_ user: SVUser, completion: ((_ error: Error?) -> Void)?) {
		self.state = .open
		
		self.user = user
		
		self.state = .established
		
		completion?(nil)
	}
	
	func sendMessage(_ message: SignalingMessage, withSessionDetails: SessionDetails?, toUser user: SVUser, completion: ((_ error: Error?) -> Void)?) {
		let error: NSError? = shouldSensMessagesSuccessfully ? nil : NSError(domain: "", code: -1, userInfo: nil)
		completion?(error)
	}
	
	func disconnectWithCompletion(_ completion: ((_ error: Error?) -> Void)?) {
		self.state = .closed
		completion?(nil)
	}
	
	var isConnected: Bool {
		return true
	}
	
	var isConnecting: Bool {
		return false
	}
}
